Added field-validation support

jn73 opened the pull request:
Generic support for form validation.

- Validator in js project
- Validations in shared project

Docs: tests and updated Validation.md in chapters-tdb.

tindzk commented on js/src/main/scala/org/widok/validation/Validator.scala:
```diff
@@ -0,0 +1,28 @@
+      channel.distinct.tail.attach(input => validateValue(channel, fieldValidators, input))
+  }
+
+  private def validateValue(channel: ReadStateChannel[_], fieldValidators: Seq[FieldValidation[_]], input: Any) = {
```
I'd suggest to remove the ``=``, to make it obvious that the method doesn't return anything.

Would it be able to make the types more specific? Perhaps something along the lines:

```scala
private def validateValue[T](channel: ReadStateChannel[T], fieldValidators: Seq[FieldValidation[T]], input: T)
```

jn73 replied:
The equal sign is gone.

I've been trying (unsuccessfully) to use a type parameter. The problem is that the validationSources needs to have en "open type" since the channels type is unknown. That's the reason why I asked if the channels for input-fields could be of any other types then strings (date, numbers, ..). But even if one could assume that all input field channels are Strings (to remove the type all together), there is the case with combined fields (password check) etc.

So, i don't see how to add a type to validateValue when it's called from the validationSources.foreach for example?

tindzk commented on docs/chapters-tbd/Validation.md:
```diff
@@ -1,2 +1,63 @@
+  val username = Var("")
+  val password = Var("")
+  val passwordVerify = Var("")
+  val samePasswords = password.combine(passwordVerify).cache("", "")
```
It should read ``cache(("", ""))``.

jn73 replied:
Ok

tindzk commented on docs/chapters-tbd/Validation.md:
```diff
@@ -1,2 +1,63 @@
+      //       Is it a bug? Shouldn't the rendered error message be removed automatically in this case?
+      //       The "fix" in this case is to add a .show(...) statement that will only show the error message when there
+      //       is a reported error for the field.
+      validator.errors.value(username).map(e => div("Validation errror: " + e.mkString(", ")).show(validator.errors.value(username).isDefined)),
```
Change this as explained in #64.

jn73 replied:
will do

tindzk replied:
That's possible already. ``ch.flatMapBuf(Buffer.from)`` should do it.

tindzk commented on docs/chapters-tbd/Validation.md:
```diff
@@ -1,2 +1,63 @@
+
+      // set "invalid" css class on password fields when validation fails
+      label("Password").forId("password"),
+      text().id("password").bind(password).cssState(validator.validations.value(password).isDefined, "invalid"),
```
This is quite long. I solved this in the Bootstrap bindings by providing an implicit, but you could just provide a helper function, so that we could write:

```scala
cssState(validator.invalid(password), "invalid")
```

jn73 replied:
Yes, i have that kind of code in my own bindings for the materializecss framework, but i wanted to use something generic in this example. I will make a helper function as you suggested.

tindzk replied:
I see. If you would like to share your Materalize bindings, we could even make them part of Widok, so that other people could benefit from them too.

tindzk commented on shared/src/main/scala/org/widok/validation/Validation.scala:
```diff
@@ -0,0 +1,142 @@
+   * @param message Error message. Variables: #{value} the validated input, #{min} = the min value
+   */
+  case class MinValidation(min: Int, message: Option[String] = None)
+    extends TextFieldValidation(Map("min" -> min), message.getOrElse("Value must be at least #{min}")) {
```
I'd prefer to use Scala's string formatting here. If it's not available in Scala.js, then the formatting logic should be moved to a separate class, so that it can be reused.

jn73 replied:
Is there a way to do this kind of formatting in scala? I've been googling but didn't come up with anything.

I need something like this (example code, won't work ofcourse):

```scala
val m = Map("car" -> "saab", "color" -> "yellow")
"My $car is $color".format(m)
```

jn73 replied:
Just to clearify. The idea is that each Validation defines a Map of specific variables for that Validation that can be used when formatting error messages #{value} is the only one that is available in all Validations. So, the formatting will take place in the ```FieldValidation```class, using the Map of variables and message that is passed in through the constructor in line 47. The actual interpolation code is a bit of a mess but I didn't find any clean way of doing this. 

```scala
  def formatMessage(msg: String, params: Map[String, Any], value: T) = {
    val paramsWithValue = params + ("value" -> value)
    (msg /: paramsWithValue) { (res, entry) => res.replaceAll("#\\{%s\\}".format(entry._1), entry._2.toString.replace( """\""", """\\""")) }
  }
```
http://stackoverflow.com/questions/4051308/better-string-formatting-in-scala

This allows the application to customize error messages, like this:

```scala
new MinValidation(10, "You entered %{value} but the min acceptable number is: %{min}")
```

tindzk replied:
If we could ensure that the parameter order stays the same throughout all messages, then we could use ``format()`` for simplicity.

tindzk commented on shared/src/main/scala/org/widok/validation/Validation.scala:
```diff
@@ -0,0 +1,142 @@
+   * @param message Error message. Variables: #{value} the validated input, #{max} tha max value
+   */
+  case class MaxValidation(max: Int, message: Option[String] = None)
+    extends TextValidation(Map("max" -> max), message.getOrElse("Value must be at most #{min}")) {
```
This should be ``#{max}``, right?

tindzk commented on shared/src/main/scala/org/widok/validation/Validation.scala:
```diff
@@ -0,0 +1,142 @@
+   * @param message Error message. Variables: #{value} the validated input
+   */
+  case class EmailValidation(message: Option[String] = None)
+    extends TextValidation(Map.empty, message.getOrElse("Must be a valid email-adress")) {
```
It should be spelt ``e-mail address``.

tindzk commented:
Two minor things left. Can be merged afterwards if nobody objects.

jn73 commented:
I have one more commit incoming. I added a method to Validator (combinedErrors) that returns a buffer of all generated error messages for a set of channels. I will add these fixes to that commit.

I also updated the documentation, but that will need some more work.
